Feature/crashloop

konst-sauce opened the pull request:
Please review the current version of a Crash Loop Detector.
This request doesn't include a separate file fix to work with Unity+Cocoa environment.

konst-sauce committed:
Added BacktraceCrashLoopDetector

konst-sauce committed:
Added public methods to configure safe mode and access its state. Removed shared instance of BCLD

konst-sauce committed:
Added detection to iOS example. Moved detection to isSafeToLaunch

konst-sauce committed:
Added Unit Tests, added logs, polished CL detector

konst-sauce committed:
Removed PLCR dependency from BTCL detector, added UI, polishes

konst-sauce committed:
Polished naming, added reset of CL detector when safe mode is necessary

konst-sauce committed:
Polished crash loop detection

Added removing of crash report when safe mode detected to prevent entering into crash loop detection loop.
Made bad events calculation reverse - from newest event back to closest successful.

konst-sauce committed:
Updated tvOS and macOS examples

vlussenburg commented on Tests/BacktraceCrashLoopDetectorTests.swift:
```diff
@@ -0,0 +1,52 @@
+            let crashLoopDetector = BacktraceCrashLoopDetector()
+            let eventsCount = BacktraceCrashLoopDetector.consecutiveCrashesThreshold
+            let timeIntervalStep = 200
+            var isCrashLoop = false
```
Why define it here? I would just keep it local to the tests. There's no reason to persist across test.

vlussenburg commented on Sources/Public/BacktraceCrashLoopDetector.swift:
```diff
@@ -0,0 +1,138 @@
+        var isSuccessful: Bool
+    }
+    
+    @objc private static let plistKey = "CrashLoopDB"
```
I don't like this name, since the file contains more than CrashLoops. It contains successful runs as well.

konst-sauce replied:
Changed to "CrashLoopDetectorData"

vlussenburg commented on Sources/Public/BacktraceCrashLoopDetector.swift:
```diff
@@ -0,0 +1,138 @@
+        return reportFullPath
+    }
+    
+    @objc private func hasCrashReport() -> Bool {
```
This doesn't address OutOfMemory crashes, since this won't leave a `live_report.plcrash`. See BacktraceOomWatcher for which files that watcher writes.

konst-sauce replied:
Added to Trello

konst-sauce committed:
Updated with comments from the pull request review

konraddysput commented on Sources/Public/BacktraceCrashLoopDetector.swift:
```diff
@@ -0,0 +1,141 @@
+        let paths = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)
+        let cacheDir = URL(fileURLWithPath: paths.isEmpty ? "" : paths[0])
+
+        let bundleIDPLCR = "com.plausiblelabs.crashreporter.data"
```
it isn't different for old plcrashreporter version and the one that Microsoft owns?

konst-sauce replied:
yes, but it is outdated now - because we rely on counter file

konraddysput commented on Sources/Public/BacktraceCrashLoopDetector.swift:
```diff
@@ -0,0 +1,141 @@
+        return exists
+    }
+
+    @objc private func addCurrentEvent() {
```
what is current and why it is important to point out this is a current event? It is always a current event no matter what happen :P  Maybe renaming it to add or addStartupEvent will make more sense here?

konst-sauce replied:
hm - basically I agree - name is a bit confusing - polished out

konraddysput commented on Sources/Public/BacktraceCrashLoopDetector.swift:
```diff
@@ -0,0 +1,141 @@
+        BacktraceLogger.debug("Event Added: \(startupEvents.count)")
+    }
+
+    @objc private func badEventsCount() -> Int {
```
badEventsCount doesn't describe also what it does - right now we're counting crashes. How about renaming it to startupCrashCounter() or something like that?

konraddysput replied:
we can also pass here as a parameter crashCounter and ask this function to check if we are in the crash loop. By doing that we do not need to iterate over the whole array (it's still small anyway and it doesn't cost as anyway, but the flow might be cleaner?)

konst-sauce replied:
1. Agree, changed it to 'consecutiveEventsCount', as crashes should be consecutive to detect crash loop.
2. The function itself calculates only how many consecutive events happened during last N (threshold) launches. Detection of a crash loop is done inside detectCrashLoop() (it loads events, adds current event, saves events, calculates consecutive, compares them with threshold and returns true or false. 
3. I will elaborate documentation and comments a bit, so it will be less confusing :P

konraddysput commented on Sources/Public/BacktraceCrashLoopDetector.swift:
```diff
@@ -0,0 +1,138 @@
+        return result
+    }
+    
+    @objc func deleteCrashReport() {
```
where do we use it?

konst-sauce replied:
marked as obsolete for now

konraddysput commented on Sources/Public/BacktraceCrashLoopDetector.swift:
```diff
@@ -0,0 +1,138 @@
+            if event.isSuccessful {
+                break
+            }
+            badEventsCount += 1
```
we shouldn't test private functions imo

konst-sauce committed:
Added Crash Loop counter

konst-sauce committed:
Polished Crash Loop detection

Added Crash Loop Counter - which counts crash loop events directly in the PLCrashReporter's signals handler.
Updated Crash Loop Detector to work with Crash Loop Counter.
Updated files structure.
Polished comments and code.

konst-sauce commented on Examples/Example-iOS/AppDelegate.swift:
```diff
@@ -18,6 +18,19 @@ final class AppDelegate: UIResponder, UIApplicationDelegate {
     func application(_ application: UIApplication,
                      didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+        
+        // Enable crash loop detector, it crashes count threshold is not specified - default will be used
```
updated

konst-sauce committed:
Updated crash loop detector algorithm, polished names, added comments.

konst-sauce committed:
Final polishes - updated CLD algorithm, updated all examples to utilise CLD functionality

konraddysput commented on Sources/Public/Internal/CrashLoopDetector/BacktraceCrashLoopCounter.swift:
```diff
@@ -0,0 +1,79 @@
+
+import Foundation
+
+@objc internal class BacktraceCrashLoopCounter: BacktraceCrashLoop {
```
1. Do we reset the counter somewhere? Let's pretend we have 4 random crashes in the last month. Do we keep 4 as a counter all the time and we wait until the 5th crash happens?
2. What we do in a situation when the session ends with a crash but it wasn't a startup crash? Its a really popular use case - In this situation we have something that we can call "crash loop" from the code perspective, however, the developer shouldn't worry about this situation.

konst-sauce replied:
1. For now counter resets when we detect a crash loop. Yes we keep 4 as a counter all the time. 
2. I didn't come with idea how to differ startup crashes right upon launch and crash somewhere after some time after launch. But comparing timestamps between didFinishLaunch and PLCR's crash signal handler maybe (though again - we have limited time here and we'll need to persist result - what brings back necessity to make a file write operation)

konst-sauce replied:
There Is no counter any more

konraddysput commented on Examples/Example-macOS-ObjC/Base.lproj/Main.storyboard:
```diff
@@ -710,16 +711,16 @@
                                 </buttonCell>
                                 <connections>
-                                    <action selector="liveReportButtonAction:" target="XfG-lQ-9wD" id="r4X-yE-pnP"/>
+                                    <action selector="crashAction:" target="XfG-lQ-9wD" id="OfJ-Ce-bpL"/>
```
shouldn't liveReportButton stay here?

konraddysput commented on Examples/Example-iOS-ObjC/AppDelegate.m:
```diff
@@ -10,6 +10,21 @@ @interface AppDelegate () <BacktraceClientDelegate>
 - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
+    
+    
+    /*  Enable crash loop detector.
```
Wouldn't be better if we describe the story what someone want to use it?

konraddysput commented on Examples/Example-iOS-ObjC/Info.plist:
```diff
@@ -15,9 +15,9 @@
 	<string>APPL</string>
 	<key>CFBundleShortVersionString</key>
-	<string>$(MARKETING_VERSION)</string>
+	<string>1.0</string>
```
is this something that we should replace with 1.0 version? I think they're overridden by cicd?

konst-sauce replied:
Restored. Changed coz Simulator was shouting about bundle version.

vlussenburg replied:
Don’t do this :) read the original commit statement when I added this. Should not be a hardcoded version. Make sure you’re on a recent xcode version. Leave it on MARKETING_VERSION.

konraddysput commented on Sources/Public/Internal/CrashLoopDetector/BacktraceCrashLoopDetector.swift:
```diff
@@ -0,0 +1,180 @@
+            false -> crash loop NOT detected -> set normal mode
+         */
+        let result = consecutiveCrashesCount >= BacktraceCrashLoopDetector.consecutiveCrashesThreshold
+        BacktraceCrashLoop.LogDebug("Finishing Crash Loop Detection: \(result)")
```
It is not important from the user perspective, maybe instead we can print something like "Is in the crash loop? \(result)" ?

konst-sauce replied:
modified

konraddysput commented on Sources/Public/Internal/CrashLoopDetector/BacktraceCrashLoopDetector.swift:
```diff
@@ -0,0 +1,180 @@
+            for not to use PLCrashReporter's APIs at all
+         */
+        
+        let bundleIDBT = Bundle.main.bundleIdentifier ?? ""
```
small nit:

backtraceBundleId would be better I believe

konst-sauce replied:
It fully duplicates PLCrashReporter's path creation, which uses main bundle, and crash report is stored under main bundle ID, not backtraces'

konraddysput commented on Sources/Public/Internal/CrashLoopDetector/BacktraceCrashLoopDetector.swift:
```diff
@@ -0,0 +1,180 @@
+extension BacktraceCrashLoopDetector {
+
+    @available(*, deprecated, message: "Temporarily not needed")
+    @objc internal func reportFilePath() -> String {
```
have you consider a situation when the crash report is from the previous session because we couldn't send it? To reproduce this problem disable internet on your phone and restart application twice. Would we increase the counter?

konst-sauce replied:
That's why we have a counter, counter incremented every time crash happens and compared with previous value from user defaults upon next launch.

konst-sauce replied:
counter increase only when crash happens, not when launch happens.
But - I'm reconsidering idea of counter - if we'll be checking crash file creation date at launch - counter wont be necessary.

konst-sauce replied:
In new logic - if file has same timestamp - it won't be considered as a crash loop continuation.

konraddysput commented on Sources/Public/Internal/CrashLoopDetector/BacktraceCrashLoopDetector.swift:
```diff
@@ -0,0 +1,180 @@
+    
+    @objc internal func consecutiveEventsCount() -> Int {
+        
+        var count = 0
```
The addStartupEvent drops the first event if the number of events is greater than the counter. You might have a situation when you will have the following number of events with crashCount property
[4,5,1,2,3] for threshold 5

let's see what is the result of the algorithm for the example above
4, 5 is going to increase counter to 2
then 1,2,3 won't increase the counter.

so as you can see the counter should be equal to 3, but is equal to 2 instead. Not sure if I'm missing something but - there is a high chance to have this situation in production code. ALso the question related to the way how we increase counter and when we reset it is really relevant to this question

konst-sauce commented on Sources/Public/Internal/CrashLoopDetector/BacktraceCrashLoopCounter.swift:
```diff
@@ -0,0 +1,79 @@
+    static internal func start() {
+        checkFileExists()
+        BacktraceCrashLoop.LogDebug("Cache Dir: \(cacheDir())")
+        BacktraceCrashLoop.LogDebug("Crash Loop Counter File Path: \(filePath())")
```
agree, removed

konst-sauce commented on Sources/Public/BacktraceClient.swift:
```diff
@@ -85,10 +94,62 @@ import Foundation
+    @objc public static func enableCrashLoopDetection(_ threshold: Int = 0) {
+        BacktraceCrashLoopCounter.start()
+
+        crashLoopDetector = BacktraceCrashLoopDetector()
```
1. Agree with passing to constructor - done.
2. We need it only at launch and using it only once, so there is no sense to keep it globally throughout the whole app and throughout the whole app's lifecycle - we don't know if final user counts every byte of memory, so using it as a local pointer just until necessary is a better way here, IMHO

konraddysput replied:
`We need it only at launch and using it only once` - so it means if someone create one more class instance, we would have them two.

konst-sauce replied:
wait -     private static var crashLoopDetector: BacktraceCrashLoopDetector? - it is static already and private, so it is like a singleton inside BacktraceClient

konraddysput replied:
you're creating a new class instance here. What will happen if you call it twice? I think you will override it. singleton would be when you check if the instance exists and if does exist then you're reusing it.

konst-sauce commented on Sources/Public/Internal/CrashLoopDetector/BacktraceCrashLoopCounter.swift:
```diff
@@ -0,0 +1,79 @@
+
+    static private func checkFileExists() {
+        let path = filePath()
+        /*
```
Added it so ...fileExists check won't be removed accidentally or intentionally.

konst-sauce commented on Sources/Public/Internal/CrashLoopDetector/BacktraceCrashLoopCounter.swift:
```diff
@@ -0,0 +1,79 @@
+
+    static private func saveCounter() {
+        let contents = String(counter)
+        try? contents.write(toFile: filePath(), atomically: true, encoding: .utf8)
```
You are right - but I have even better idea - remove writing counter to the file at all here - we can check PLCR's report file's creation data on next launch, but not sure how is it reliable - will do some tests.

konraddysput replied:
how you know if the crash is from the previous session? you could have a situation when we couldn't send a crash and the crash is still available in the crash dir

konst-sauce replied:
counter removed completely

konst-sauce committed:
Polished for comments in PR

konst-sauce committed:
backtrace.version attribute fix

konst-sauce committed:
CLD algorithm polished. Some PR comments utilised.

konst-sauce committed:
Polished other example apps, fixed bug.
